### PR TITLE
[CDX-217] Add Sale Price Conditional Check

### PR DESCRIPTION
**Pull request by evanyan13**
### Objective
* As user of PLP library I want to add a conditional check to safeguard against situations where we are mistakenly passing the same value to both `sale_price` and `price`
* As CDX, I want to implement a logic to only render `sale_price` when it is strictly less than `price`

### Definition of Done
* Add conditional check to only render `sale_price` and its UI when it is strictly less than `price`
* Add tests

### CODE REVIEW

File: spec/test-utils.tsx [hunk 1]
```diff
@@ -51,11 +51,32 @@ const delay = (ms) =>
 const getAttribute = (container) => (attribute) =>
   container.querySelector(`[${attribute}]`)?.getAttribute(attribute);
 
+function transformSalePrice(item, mockSalePrice) {
+  const itemCopy = { ...item };
+  if (itemCopy.variations && Array.isArray(itemCopy.variations)) {
+    itemCopy.variations = itemCopy.variations.map((variation) => ({
```
Comment (TarekAlQaddy) [inline spec/test-utils.tsx:57]: I'm curios why you're copying the item and the variations to add the `sale_price` attribute
How about iterating through the `variations` and just assigning the mockSalePrice inside?
Something like
```
itemCopy.variations.forEach((variation) => {
  variation.data.sale_price = mockSalePrice
})
```

If the intention of that function is to also copy the item as to not mess with the original data, I guess we should suggest a better naming? what do you think?

File: src/hooks/useProduct.ts [hunk 2]
```diff
@@ -20,12 +20,17 @@ const useProductInfo: UseProductInfo = ({ item }) => {
...
   const itemName = productSwatch?.selectedVariation?.itemName || item.itemName;
   const itemPrice = productSwatch?.selectedVariation?.price || getPrice(item);
-  const salePrice = productSwatch?.selectedVariation?.salePrice || getSalePrice(item);
   const itemImageUrl = productSwatch?.selectedVariation?.imageUrl || item.imageUrl;
   const itemUrl = productSwatch?.selectedVariation?.url || item.url;
   const variationId = productSwatch?.selectedVariation?.variationId;
   const { itemId } = item;
 
+  function isValidSalePrice(salePrice: number, usualPrice: number) {
+    return salePrice && usualPrice && salePrice >= 0 && salePrice < usualPrice;
```
Comment (TarekAlQaddy) [inline src/hooks/useProduct.ts:29]: I can see `0` is included in `salePrice >= 0` but this will never be reached if the sale price is actually `0` because the first part of the expression will be falsy

**Review by TarekAlQaddy — CHANGES_REQUESTED** [review]
Thanks @evanyan13 for working on that!
I left some minor comments, please let me know what you think

File: src/hooks/useProduct.ts (hunk 2 above)
Comment (evanyan13) [reply src/hooks/useProduct.ts:29]: Hi @TarekAlQaddy 
I think we should keep this as `negative` numbers are considered "truthy".
And it might return negative prices as valid if we don't include an explicit check here

Comment (TarekAlQaddy) [reply src/hooks/useProduct.ts:29]: @evanyan13 correct, but I was referring to the `0` value. I you want to consider the case where the sale price is 0 (I can see it's included in `salePrice >= 0`) then you should also consider it in the first part of the expression as `salePrice && ...` will evaluate to `false` in case `salePrice` is `0`

Comment (evanyan13) [reply src/hooks/useProduct.ts:29]: @TarekAlQaddy 
Ahh my bad, I just understood what you mean.
This might be a question of whether we want to display sale_price that are 0 or not?
If we don't want to display 0 as sale_price, we can just update the logic:
Instead of
```
salePrice >= 0
```
We just use
```
salePrice > 0
```
So `salePrice && ...` will evaluate to `false` in case when `salePrice` is `0`

Comment (TarekAlQaddy) [reply src/hooks/useProduct.ts]: @evanyan13 Thanks for making the changes and sorry for the back and forth but looks like there are more places that we need to check in order to handle the `0` price, I just tested it and it didn't show. I checked why and seems that you should handle `0` in the productCard [here](https://github.com/Constructor-io/constructorio-ui-plp/blob/main/src/components/ProductCard/ProductCard.tsx#L73) and in the price formatter [here](https://github.com/Constructor-io/constructorio-ui-plp/blob/main/src/utils/formatters.ts#L3)

Comment (evanyan13) [reply src/hooks/useProduct.ts]: Thank you @TarekAlQaddy 
Just to note here that based on our discussion, we have decided to not support `0` as a valid sale price value as there are other existing logic that blocks `0` from being rendered as mentioned above.
And we should not add more logic just to handle this edge case

File: src/hooks/useProduct.ts [hunk 3]
```diff
@@ -20,12 +20,17 @@ const useProductInfo: UseProductInfo = ({ item }) => {
 
   const itemName = productSwatch?.selectedVariation?.itemName || item.itemName;
   const itemPrice = productSwatch?.selectedVariation?.price || getPrice(item);
-  const salePrice = productSwatch?.selectedVariation?.salePrice || getSalePrice(item);
   const itemImageUrl = productSwatch?.selectedVariation?.imageUrl || item.imageUrl;
   const itemUrl = productSwatch?.selectedVariation?.url || item.url;
   const variationId = productSwatch?.selectedVariation?.variationId;
   const { itemId } = item;
 
+  function isValidSalePrice(salePrice: number, usualPrice: number) {
```
Comment (TarekAlQaddy) [inline src/hooks/useProduct.ts:28]: a small performance tweak: Would it be possible to move this function definition outside the hook? maybe above it or in the utils
Since it doesn't have any side effects and is not dependent on any of the hook's internals we can avoid defining the function in each run of the hook and only define it once

**Review by TarekAlQaddy — APPROVED** [review]
Approved with a minor comment, please let me know what you think

File: src/hooks/useProduct.ts (hunk 3 above)
Comment (evanyan13) [reply src/hooks/useProduct.ts:28]: I see I get what you mean here
Maybe we can place it in the `/utils/common`? I can't seem to find suitable place for this actually

File: src/components/ProductCard/ProductCard.tsx [hunk 4]
```diff
@@ -70,7 +70,7 @@ export default function ProductCard(props: ProductCardProps) {
   const onAddToCart = useOnAddToCart(client, state.callbacks.onAddToCart);
   const { formatPrice } = state.formatters;
   const onClick = useOnProductCardClick(client, state.callbacks.onProductCardClick);
-  const hasSalesPrice = useMemo(() => !!(salePrice && Number(salePrice) >= 0), [salePrice]);
+  const hasSalesPrice = useMemo(() => !!salePrice, [salePrice]);
```
Comment (Mudaafi) [inline src/components/ProductCard/ProductCard.tsx:73]: @evanyan13 Sale Price could reasonably be 0 right? Wouldn't this fail that edge case?

File: src/hooks/useProduct.ts [hunk 5]
```diff
@@ -20,12 +20,17 @@ const useProductInfo: UseProductInfo = ({ item }) => {
...
+    return salePrice && usualPrice && salePrice > 0 && salePrice < usualPrice;
+  }
+  let salePrice = productSwatch?.selectedVariation?.salePrice || getSalePrice(item);
```
Comment (Mudaafi) [inline src/hooks/useProduct.ts:31]: small lint thing: newline after function declaration
```suggestion
  }
  
  let salePrice = productSwatch?.selectedVariation?.salePrice || getSalePrice(item);
```

File: src/hooks/useProduct.ts [hunk 6]
```diff
@@ -20,12 +20,17 @@ const useProductInfo: UseProductInfo = ({ item }) => {
...
   const itemName = productSwatch?.selectedVariation?.itemName || item.itemName;
   const itemPrice = productSwatch?.selectedVariation?.price || getPrice(item);
-  const salePrice = productSwatch?.selectedVariation?.salePrice || getSalePrice(item);
   const itemImageUrl = productSwatch?.selectedVariation?.imageUrl || item.imageUrl;
   const itemUrl = productSwatch?.selectedVariation?.url || item.url;
   const variationId = productSwatch?.selectedVariation?.variationId;
   const { itemId } = item;
 
+  function isValidSalePrice(salePrice: number, usualPrice: number) {
+    return salePrice && usualPrice && salePrice > 0 && salePrice < usualPrice;
```
Comment (Mudaafi) [inline src/hooks/useProduct.ts:29]: Same edge case as above, salePrice can reasonably be $0 right?

File: src/components/ProductCard/ProductCard.tsx (hunk 4 above)
Comment (Mudaafi) [reply src/components/ProductCard/ProductCard.tsx:73]: Instead of doing this in the component, let's just return an `isOnSale` field in the `useProductInfo` hook and consume it here. Doesn't make sense to do the logical check twice imo.

File: spec/hooks/useProductInfo/useProductInfo.test.js [hunk 7]
```diff
@@ -120,4 +95,63 @@ describe('Testing Hook: useProductInfo', () => {
...
+    });
+
+    it('Should return salePrice if salePrice is valid (positive and less than price)', async () => {
+      const item = transformResultItem(copyItemWithNewSalePrice(mockItemWithSalePrice, 1));
+      const { result } = renderHookWithCioPlp(() => useProductInfo({ item }));
+      await waitFor(() => {
+        expect(result.current.salePrice).toEqual(1);
+      });
+    });
+  });
```
Comment (Mudaafi) [inline spec/hooks/useProductInfo/useProductInfo.test.js:156]: How do you feel if we reformat this to follow the `.each` pattern above? I know the test itself isn't bulky, but there's basically three lines we keep repeating for each unit test.

**Review by Mudaafi — CHANGES_REQUESTED** [review]
Thanks for making the PR! A small edge case consideration before we ship this.

File: src/components/ProductCard/ProductCard.tsx (hunk 4 above)
Comment (Mudaafi) [reply src/components/ProductCard/ProductCard.tsx]: ah I missed that discussion. Sounds good, we can take this as a future tech-debt story to look into since it touches multiple parts. Mind creating that story so we don't forget?
